### PR TITLE
Enable ORANGE demo loop

### LINKED ISSUE #267: Implement portable in-house geometry
There are several reasons to support an alternative geometry implementation to VecGeom:
- Exploring performance portability (VecGeom is very closely tied to CUDA)
- Evaluating how the app's overall performance changes when using a less expensive geometry implementation
- Reducing build/CI times

To do this we'll want to:
- Refine geometry propagation/tracking interface. Look at the current use cases for the geometry track view interface, simplify and/or document which functions are part of the "public" interface.
- Possibly template key parts of the code on geometry type, and/or use virtual geometry interfaces for other parts of the CPU code (such as cell labels and high-level kernel launch functionality)

Then we'll add a new surrogate geometry implementation (possibly based on RTK from Shift) for simple geometries with planes and cylinders. Maybe define a new JSON (rather than Teuchos XML) for on-disk representation.

ORANGE:
- [x] Surface intersections (#291)
- [x] Static polymorphism and operator (#306)
- [x] JSON I/O (#307)
- [x] Volumes and initialization (#312)
- [x] Distance-to-boundary (#335)
- [x] Params and TrackView interface (#336)
- [x] Compile-time switching between VecGeom and ORANGE for main codebase (#337)
- [x] Enable previously vecgeom-only tests (#339 )
- [ ] Enable previously vecgeom-only demo (#343)

### PR DESCRIPTION
**Pull request by sethrj**
- Fix (and add tests for) missing write-on-destruct for ORANGE track view
- Fix material mapping in demo loop
- Update CMS model to remove pointer addresses, and add ORANGE JSON input
- Use JSON file instead of GDML if VecGeom is disabled

Closes #267

### CODE REVIEW

File: app/demo-loop/simple-cms.org.json [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+{
+"faces": [
+7,
+8,
+11,
+12
+],
+"flags": 1,
+"logic": "0 1 ~ & 2 ~ & ~ 0 1 ~ & 3 ~ & &",
+"num_intersections": 6,
```
Comment (paulromano) [inline app/demo-loop/simple-cms.org.json:119]: Just curious -- in general, how is the number of intersections determined?

File: app/demo-loop/simple-driver.py [hunk 2]
```diff
@@ -5,16 +5,16 @@
...
+from distutils.util import strtobool
 from os import environ, path
 from sys import exit, argv
 
 try:
-    geometry_filename = argv[1]
-    hepmc3_filename = argv[2]
-except (IndexError, TypeError):
+    (geometry_filename, hepmc3_filename) = argv[1:]
+except IndexError:
```
[the remark this answers is not in the document]
Comment (sethrj) [reply app/demo-loop/simple-driver.py:17]: Ah damn, I thought incorrect tuple unpacking was IndexError... I'll replace with ValueError.

File: test/orange/Orange.test.cc [hunk 3]
```diff
@@ -179,4 +179,26 @@ TEST_F(TwoVolumeTest, persistence)
...
+        EXPECT_EQ(VolumeId{0}, geo.volume_id());
+        EXPECT_EQ(SurfaceId{0}, geo.surface_id());
+        EXPECT_VEC_SOFT_EQ(Real3({-1.5, 0, 0}), geo.pos());
+        geo.set_dir({1, 0, 0});
+    }
+    {
+        auto geo = this->make_track_view();
+        EXPECT_VEC_SOFT_EQ(Real3({1, 0, 0}), geo.dir());
+        geo.find_next_step();
+        geo.move_across_boundary();
```
[the remark this answers is not in the document]
Comment (sethrj) [reply test/orange/Orange.test.cc:193]: Ugh, good catch: it's probably moving all the way to the far side of the sphere (skipping the zero-distance surface), but I don't actually check. I'll fix it.

File: test/orange/Orange.test.cc [hunk 4]
```diff
@@ -179,4 +179,26 @@ TEST_F(TwoVolumeTest, persistence)
...
+    }
+    {
+        auto geo = this->make_track_view();
+        EXPECT_VEC_SOFT_EQ(Real3({-1.5, .5, .5}), geo.pos());
+        geo.set_dir({1, 0, 0});
+    }
+    {
+        auto geo = this->make_track_view();
+        EXPECT_VEC_SOFT_EQ(Real3({1, 0, 0}), geo.dir());
+        EXPECT_SOFT_EQ(3.0, geo.find_next_step());
```
Comment (paulromano) [inline test/orange/Orange.test.cc:197]: Now I'm confused about this one. If the track is at (-1.5, 0.5, 0.5) heading to the right, shouldn't it hit the sphere?

Comment (paulromano) [reply test/orange/Orange.test.cc:197]: Ok, I think I see what's going on. When we call `move_internal`, it doesn't invalidate/change `local_.surface`, so when we go to calculate intersections for the sphere, it still thinks it's on the surface even though it's not.

Comment (sethrj) [reply test/orange/Orange.test.cc:197]: Damn, another good catch: that should reset the surface. I clearly didn't test that function enough 🤦

**Review by paulromano — APPROVED** [review]
All looks good now. Thanks @sethrj!

There's a CI failure now but looks like some kind of system issue...

File: app/demo-loop/simple-cms.org.json (hunk 1 above)
Comment (sethrj) [reply app/demo-loop/simple-cms.org.json:119]: It's the sum of the number of intersection per surfaces over all surfaces. So the inner and outer cylinder each have two, and the Z caps each have one, for a total of 6.

Comment (paulromano) [reply app/demo-loop/simple-cms.org.json:119]: Well, that seems obvious in retrospect now :blush:

**Pull request merged**